File: backend/app/Models/ProductoPresentacion.php
<?php
namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class ProductoPresentacion extends Model
{
    protected $table = 'producto_presentaciones';

    protected $fillable = [
        'producto_id',
        'nombre',
        'codigo_barras',
        'precio_venta',
        'precio_compra',
        'margen',
        'factor_conversion',
        'unidad_base_id',
        'producto_complementario_id',
        'cantidad_complementaria',
        'activo',
    ];

    protected function casts(): array
    {
        return [
            'precio_venta' => 'decimal:2',
            'precio_compra' => 'decimal:2',
            'margen' => 'decimal:2',
            'factor_conversion' => 'decimal:3',
            'cantidad_complementaria' => 'decimal:2',
            'activo' => 'boolean',
        ];
    }

    public function producto() { return $this->belongsTo(Producto::class); }
    public function unidadBase() { return $this->belongsTo(UnidadMedida::class, 'unidad_base_id'); }
    public function complementario() { return $this->belongsTo(Producto::class, 'producto_complementario_id'); }
}

File: app_abarrotes/lib/models/producto_presentacion.dart
class ProductoPresentacion {
  final int id;
  final int productoId;
  final String nombre;
  final String? codigoBarras;
  final double precioVenta;
  final double factorConversion;
  final int? unidadBaseId;
  final String? unidadBaseNombre;
  final bool activo;

  ProductoPresentacion({
    required this.id,
    required this.productoId,
    required this.nombre,
    this.codigoBarras,
    this.precioVenta = 0,
    this.factorConversion = 1,
    this.unidadBaseId,
    this.unidadBaseNombre,
    this.activo = true,
  });

  factory ProductoPresentacion.fromJson(Map<String, dynamic> json) {
    return ProductoPresentacion(
      id: json['id'],
      productoId: json['producto_id'],
      nombre: json['nombre'],
      codigoBarras: json['codigo_barras'],
      precioVenta: (json['precio_venta'] ?? 0).toDouble(),
      factorConversion: (json['factor_conversion'] ?? 1).toDouble(),
      unidadBaseId: json['unidad_base']?['id'],
      unidadBaseNombre: json['unidad_base']?['nombre'],
      activo: json['activo'] ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
    'nombre': nombre,
    'codigo_barras': codigoBarras,
    'precio_venta': precioVenta,
    'factor_conversion': factorConversion,
    'unidad_base_id': unidadBaseId,
    'activo': activo,
  };
}
